File: dotnet/src/Proc.cs
// Subprocess execution with timeouts (stands in for execFile/spawn/spawnSync).
// Arguments are passed as an argv list, never through a shell.

using System.Diagnostics;
using System.Text;

namespace Skills;

internal sealed record ProcOutput(int? Status, byte[] Stdout, byte[] Stderr)
{
    public bool Success => Status == 0;
    public string StdoutText => Encoding.UTF8.GetString(Stdout);
    public string StderrText => Encoding.UTF8.GetString(Stderr);
}

internal enum ProcErrorKind
{
    NotFound,
    Timeout,
    TooMuchOutput,
    Other,
}

internal sealed class ProcException(ProcErrorKind kind, string message) : Exception(message)
{
    public ProcErrorKind Kind { get; } = kind;
}

internal sealed class Proc
{
    private readonly string _program;
    private readonly List<string> _args;
    private readonly Dictionary<string, string> _env = new();
    private TimeSpan? _timeout;
    private int? _maxOutput;
    private bool _hideWindow;

    private Proc(string program, IEnumerable<string> args)
    {
        _program = program;
        _args = args.ToList();
    }

    /// Children share the parent's console, like Node's default `windowsHide: false`.
    public static Proc Command(string program, params string[] args) => new(program, args);

    public static Proc Command(string program, IEnumerable<string> args) => new(program, args);

    public Proc Env(string k, string v)
    {
        _env[k] = v;
        return this;
    }

    public Proc Timeout(TimeSpan t)
    {
        _timeout = t;
        return this;
    }

    public Proc MaxOutput(int bytes)
    {
        _maxOutput = bytes;
        return this;
    }

    /// Node's `windowsHide: true`.
    public Proc HideWindow()
    {
        _hideWindow = true;
        return this;
    }

    private ProcessStartInfo StartInfo(bool redirectStdin, bool redirectOut)
    {
        var psi = new ProcessStartInfo(_program)
        {
            UseShellExecute = false,
            RedirectStandardInput = redirectStdin,
            RedirectStandardOutput = redirectOut,
            RedirectStandardError = redirectOut,
            CreateNoWindow = _hideWindow,
        };
        foreach (var a in _args) psi.ArgumentList.Add(a);
        foreach (var (k, v) in _env) psi.Environment[k] = v;
        return psi;
    }

    private static Process Start(ProcessStartInfo psi)
    {
        try
        {
            return Process.Start(psi) ?? throw new ProcException(ProcErrorKind.Other, "failed to start process");
        }
        catch (System.ComponentModel.Win32Exception e) when (e.NativeErrorCode is 2 or 3)
        {
            throw new ProcException(ProcErrorKind.NotFound, "command not found");
        }
        catch (System.ComponentModel.Win32Exception e)
        {
            throw new ProcException(ProcErrorKind.Other, e.Message);
        }
    }

    /// Run with stdin closed and stdout/stderr captured.
    public ProcOutput Output()
    {
        using var p = Start(StartInfo(redirectStdin: true, redirectOut: true));
        p.StandardInput.Close();
        var outTask = ReadAllAsync(p.StandardOutput.BaseStream);
        var errTask = ReadAllAsync(p.StandardError.BaseStream);
        var ms = _timeout.HasValue ? (int)Math.Min(int.MaxValue, _timeout.Value.TotalMilliseconds) : -1;
        if (!p.WaitForExit(ms))
        {
            try { p.Kill(entireProcessTree: true); } catch { /* already exited */ }
            throw new ProcException(ProcErrorKind.Timeout, "timed out");
        }
        p.WaitForExit();
        var stdout = outTask.GetAwaiter().GetResult();
        var stderr = errTask.GetAwaiter().GetResult();
        if (_maxOutput.HasValue && stdout.Length + stderr.Length > _maxOutput.Value)
            throw new ProcException(ProcErrorKind.TooMuchOutput, "output exceeded buffer");
        return new ProcOutput(p.ExitCode, stdout, stderr);
    }

    /// Run with inherited stdio, returning the exit code.
    public int StatusInherit()
    {
        using var p = Start(StartInfo(redirectStdin: false, redirectOut: false));
        p.WaitForExit();
        return p.ExitCode;
    }

    /// spawnSync with `stdio: ['inherit', 'pipe', 'pipe']`.
    public ProcOutput OutputInheritStdin()
    {
        using var p = Start(StartInfo(redirectStdin: false, redirectOut: true));
        var outTask = ReadAllAsync(p.StandardOutput.BaseStream);
        var errTask = ReadAllAsync(p.StandardError.BaseStream);
        p.WaitForExit();
        return new ProcOutput(p.ExitCode, outTask.GetAwaiter().GetResult(), errTask.GetAwaiter().GetResult());
    }

    private static async Task<byte[]> ReadAllAsync(Stream s)
    {
        using var ms = new MemoryStream();
        await s.CopyToAsync(ms).ConfigureAwait(false);
        return ms.ToArray();
    }
}

File: rust/src/proc.rs
//! Subprocess execution with timeouts (stands in for `execFile`/`spawn`).

use std::io::Read;
use std::process::{Command, Stdio};
use std::time::{Duration, Instant};

pub struct Output {
    pub status: Option<i32>,
    pub stdout: Vec<u8>,
    pub stderr: Vec<u8>,
}

impl Output {
    pub fn success(&self) -> bool {
        self.status == Some(0)
    }
    pub fn stdout_str(&self) -> String {
        String::from_utf8_lossy(&self.stdout).into_owned()
    }
    pub fn stderr_str(&self) -> String {
        String::from_utf8_lossy(&self.stderr).into_owned()
    }
}

#[derive(Debug)]
pub enum ProcError {
    /// The executable was not found (ENOENT).
    NotFound,
    /// The process exceeded its timeout and was killed.
    Timeout,
    /// Output exceeded the buffer limit and the process was killed.
    TooMuchOutput,
    Other(String),
}

impl std::fmt::Display for ProcError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            ProcError::NotFound => write!(f, "command not found"),
            ProcError::Timeout => write!(f, "timed out"),
            ProcError::TooMuchOutput => write!(f, "output exceeded buffer"),
            ProcError::Other(s) => write!(f, "{}", s),
        }
    }
}

pub struct Run {
    cmd: Command,
    timeout: Option<Duration>,
    max_output: Option<usize>,
}

pub fn command(program: &str, args: &[&str]) -> Run {
    // Children share the parent's console, like Node's spawn/execFile default
    // (`windowsHide: false`). Use `hide_window` where TS passes `windowsHide`.
    let mut cmd = Command::new(program);
    cmd.args(args);
    Run {
        cmd,
        timeout: None,
        max_output: None,
    }
}

impl Run {
    pub fn env(mut self, k: &str, v: &str) -> Self {
        self.cmd.env(k, v);
        self
    }
    pub fn timeout(mut self, d: Duration) -> Self {
        self.timeout = Some(d);
        self
    }
    pub fn max_output(mut self, n: usize) -> Self {
        self.max_output = Some(n);
        self
    }
    /// Node's `windowsHide: true`: start the child without a console window.
    pub fn hide_window(mut self) -> Self {
        #[cfg(windows)]
        {
            use std::os::windows::process::CommandExt;
            const CREATE_NO_WINDOW: u32 = 0x0800_0000;
            self.cmd.creation_flags(CREATE_NO_WINDOW);
        }
        self
    }

    /// Run with stdin closed and stdout/stderr captured.
    pub fn output(mut self) -> Result<Output, ProcError> {
        self.cmd
            .stdin(Stdio::null())
            .stdout(Stdio::piped())
            .stderr(Stdio::piped());
        let mut child = match self.cmd.spawn() {
            Ok(c) => c,
            Err(e) if e.kind() == std::io::ErrorKind::NotFound => return Err(ProcError::NotFound),
            Err(e) => return Err(ProcError::Other(e.to_string())),
        };
        let mut out_pipe = child.stdout.take().unwrap();
        let mut err_pipe = child.stderr.take().unwrap();
        let out_t = std::thread::spawn(move || {
            let mut v = Vec::new();
            let _ = out_pipe.read_to_end(&mut v);
            v
        });
        let err_t = std::thread::spawn(move || {
            let mut v = Vec::new();
            let _ = err_pipe.read_to_end(&mut v);
            v
        });
        let start = Instant::now();
        let status = loop {
            match child.try_wait() {
                Ok(Some(s)) => break s,
                Ok(None) => {}
                Err(e) => return Err(ProcError::Other(e.to_string())),
            }
            if let Some(t) = self.timeout {
                if start.elapsed() > t {
                    let _ = child.kill();
                    let _ = child.wait();
                    return Err(ProcError::Timeout);
                }
            }
            std::thread::sleep(Duration::from_millis(15));
        };
        let stdout = out_t.join().unwrap_or_default();
        let stderr = err_t.join().unwrap_or_default();
        if let Some(max) = self.max_output {
            if stdout.len() + stderr.len() > max {
                return Err(ProcError::TooMuchOutput);
            }
        }
        Ok(Output {
            status: status.code(),
            stdout,
            stderr,
        })
    }

    /// Run with inherited stdio, returning the exit code.
    pub fn status_inherit(mut self) -> Result<Option<i32>, ProcError> {
        match self.cmd.status() {
            Ok(s) => Ok(s.code()),
            Err(e) if e.kind() == std::io::ErrorKind::NotFound => Err(ProcError::NotFound),
            Err(e) => Err(ProcError::Other(e.to_string())),
        }
    }

    /// Run with stdin inherited and stdout/stderr captured (spawnSync with
    /// `stdio: ['inherit', 'pipe', 'pipe']`).
    pub fn output_inherit_stdin(mut self) -> Result<Output, ProcError> {
        self.cmd
            .stdin(Stdio::inherit())
            .stdout(Stdio::piped())
            .stderr(Stdio::piped());
        match self.cmd.output() {
            Ok(o) => Ok(Output {
                status: o.status.code(),
                stdout: o.stdout,
                stderr: o.stderr,
            }),
            Err(e) if e.kind() == std::io::ErrorKind::NotFound => Err(ProcError::NotFound),
            Err(e) => Err(ProcError::Other(e.to_string())),
        }
    }
}
